### DISCUSSION TITLE
Calculate Cavg per interval in RxODE

### DISCUSSION THREAD

**Discussion by krinaj** [nlmixrdevelopment/RxODE#444] [Q&A]
Hello,

I am trying to simulate simultaneous PK PD model using RxODE. For PK, I am trying to see if I can sample every xx hr and then calculate Cavg per day to feed into PD part of the model. Here is my messy code for this in RxODE, but am running into issues. I am getting an error message state '"AUCR1': 'lag', 'lead', 'first', 'last', 'diff' not legal:". Anyways, I am not sure if this is the right way to do this. Any suggestions? 

  d/dt(AUCR1) = centR/VR ; 
  AUCprevR <- lag(AUCR1, 1) ;
  AUClagR <- AUCR1 - AUCprevR ;
  tinterval = 1 ;    ##1 day
  Cavg = AUClagR/tinterval  ;

Thanks,
Krina

**Comment by mattfidler** [comment]
Hi @krinaj 

To me this seems reasonable as long as you only have dosing records for the PK records. You would have to add ` addDosing = TRUE` to see the values in your output.  I cannot remember the parsing rules for `lag`, `lead` etc. If you post a reproducible example, I can help more with this method.

On the other hand, another option is to add a dosing record for the dose in the `AUCR1` that resets the value to zero with `evid = 5`.  

Then you would only divide this auc by the inter-dose interval (which can be retrieved with `tad(auc)`). 

An example of this using the `RxODE` tutorial model is

``` r
library(RxODE)
#> RxODE 1.1.0 using 4 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`


## Model from RxODE tutorial
m1 <-RxODE({
    KA=2.94E-01;
    CL=1.86E+01;
    V2=4.02E+01;
    Q=1.05E+01;
    V3=2.97E+02;
    Kin=1;
    Kout=1;
    EC50=200;
    ## Added modeled bioavaiblity, duration and rate
    fdepot = 1;
    durDepot = 8;
    rateDepot = 1250;
    C2 = centr/V2;
    C3 = peri/V3;
    d/dt(depot) =-KA*depot;
    f(depot) = fdepot
    d/dt(centr) = KA*depot - CL*C2 - Q*C2 + Q*C3;
    d/dt(peri)  =                    Q*C2 - Q*C3;
    d/dt(eff)  = Kin - Kout*(1-C2/(EC50+C2))*eff;
    eff(0) = 1
    # Method to calculate a running Cavg
    # First calculate the AUC for the interval
    d/dt(auc) = C2
    # Time interval for calculating Cavg
    tadAuc = tad(auc)
    # Divide by the interval (unless the interval is zero)
    Cavg = auc / ifelse(tadAuc == 0, 1, tadAuc)
})


ev <- et(timeUnits="hr") %>%
  et(amt=10000, ii=12,until=24 * 2, cmt="depot") %>%
  et(amt=0, ii=12, until=24 * 2, cmt="auc", evid=5) %>% # reset to zero with every dose
  et(seq(0, 24 * 2, length.out=1000))

s <- rxSolve(m1, ev)

plot(s, C2, auc, Cavg)
```

![](https://i.imgur.com/Mp2UUAC.png)

<sup>Created on 2021-08-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

I would hope that you would be using this with a linked model, since as you can see there are major discontinuities in the running Cavg values

**Reply by mattfidler** [reply]
The ifelse actually isn't required

See:
``` r
library(RxODE)
#> RxODE 1.1.0 using 4 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`


## Model from RxODE tutorial
m1 <-RxODE({
    KA=2.94E-01;
    CL=1.86E+01;
    V2=4.02E+01;
    Q=1.05E+01;
    V3=2.97E+02;
    Kin=1;
    Kout=1;
    EC50=200;
    ## Added modeled bioavaiblity, duration and rate
    fdepot = 1;
    durDepot = 8;
    rateDepot = 1250;
    C2 = centr/V2;
    C3 = peri/V3;
    d/dt(depot) =-KA*depot;
    f(depot) = fdepot
    d/dt(centr) = KA*depot - CL*C2 - Q*C2 + Q*C3;
    d/dt(peri)  =                    Q*C2 - Q*C3;
    d/dt(eff)  = Kin - Kout*(1-C2/(EC50+C2))*eff;
    eff(0) = 1
    # Method to calculate a running Cavg
    # First calculate the AUC for the interval
    d/dt(auc) = C2
    # Time interval for calculating Cavg
    tadAuc <- tad(auc)
    Cavg <- auc / tad(auc)
})


ev <- et(timeUnits="hr") %>%
  et(amt=10000, ii=12,until=24 * 2, cmt="depot") %>%
  et(amt=0, ii=12, until=24 * 2, cmt="auc", evid=5) %>% # reset to zero with every dose
  et(seq(0, 24 * 2, length.out=1000))

s <- rxSolve(m1, ev)

plot(s, C2, auc, Cavg, tadAuc)
```

![](https://i.imgur.com/ar2kZ9M.png)

<sup>Created on 2021-08-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Thanks for your patience waiting for this response.  If you have a chance, let me know if this works for you.

**Comment by krinaj** [comment]
This is great. Thanks, Mattew! I will try it out today.
